src/lib/Repository/NameSchema/NameSchemaService.php

```diff
@@ -0,0 +1,425 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Core\Repository\NameSchema;
+
+use Ibexa\Contracts\Core\Event\ResolveUrlAliasSchemaEvent;
+use Ibexa\Contracts\Core\Repository\NameSchema\NameSchemaServiceInterface;
+use Ibexa\Contracts\Core\Repository\NameSchema\SchemaIdentifierExtractorInterface;
+use Ibexa\Contracts\Core\Repository\Values\Content\Content;
+use Ibexa\Contracts\Core\Repository\Values\ContentType\ContentType;
+use Ibexa\Core\FieldType\FieldTypeRegistry;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @internal Meant for internal use by Repository.
+ *
+ * NameSchemaService is internal service for resolving content name and url alias patterns.
+ * This code supports content name pattern groups.
+ *
+ * Syntax:
+ * <code>
+ * &lt;attribute_identifier&gt;
+ * &lt;attribute_identifier&gt; &lt;2nd-identifier&gt;
+ * User text &lt;attribute_identifier&gt;|(&lt;2nd-identifier&gt;&lt;3rd-identifier&gt;)
+ * </code>
+ *
+ * Example:
+ * <code>
+ * &lt;nickname|(&lt;firstname&gt; &lt;lastname&gt;)&gt;
+ * </code>
+ *
+ * Tokens are looked up from left to right. If a match is found for the
+ * leftmost token, the 2nd token will not be used. Tokens are representations
+ * of fields. So a match means that the current field has data.
+ *
+ * Tokens are the field definition identifiers which are used in the class edit-interface.
+ */
+class NameSchemaService implements NameSchemaServiceInterface
+{
+    /**
+     * The string to use to signify group tokens.
+     *
+     * @var string
+     */
+    public const META_STRING = 'EZMETAGROUP_';
+
+    protected FieldTypeRegistry $fieldTypeRegistry;
+
+    /**
+     * @param array{limit?: integer, sequence?: string} $settings
+     */
+    protected array $settings;
+
+    private EventDispatcherInterface $eventDispatcher;
+
+    private SchemaIdentifierExtractorInterface $schemaIdentifierExtractor;
+
+    public function __construct(
+        FieldTypeRegistry $fieldTypeRegistry,
+        SchemaIdentifierExtractorInterface $schemaIdentifierExtractor,
+        EventDispatcherInterface $eventDispatcher,
+        array $settings = []
+    ) {
```

Comment:
The comment https://github.com/ibexa/core/pull/242#discussion_r1255665823 has not been addressed.
It was about type-hint for `\Ibexa\Core\Repository\NameSchema\NameSchemaService::__construct` `$settings` array argument 

```suggestion
   /**
    * @param array{limit?: integer, sequence?: string} $settings
    */
    public function __construct(
        FieldTypeRegistry $fieldTypeRegistry,
        SchemaIdentifierExtractorInterface $schemaIdentifierExtractor,
        EventDispatcherInterface $eventDispatcher,
        array $settings = []
    ) {
```